### ISSUE TITLE
`Color` should not have `InvalidColor`

### ISSUE THREAD

**Issue by ghivert** [brunoti/lumiere#1]
I think it's counter-intuitive to be able to get a variable of type `Color` which in fact, is not a color.

Functions like `parse_hex` should return `Result(Color, Error)`, with `Error` being something like `pub type Error { NotAColor(String) }`. Otherwise, right now, the `Result(Color, Color)` is redundant, you could completely return directly a `Color`.

This ensures we don't have to wonder if the Color is really a color or not, and we can always use `let assert Ok(clr) = parse_hex("#000")` if we're 100% sure about what we're doing.

This would also have the nice effect to remove `is_valid`, since when manipulating a `Color`, it's always a valid color. Let's handle errors before doing some computations, to remove all the `Result` from the `to_xxx` functions.

**Comment by brunoti** [comment]
Yeah, that part I'm still thinking about it because I wanted it to work like `Result` so the user can build on top of it and only handle errors in the end.

```gleam
parse_hex("#FFF")
  |> darken(50)
  |> alpha(90)
  |> to_rgba() // so here it would return Result(String, Error)
```

Maybe: 

```gleam
parse_hex("#FFF")
  |> darken(50)
  |> alpha(90)
  |> to_rgba(fallback: "#FFF")
```

You suggest that we do something like:

```gleam
let assert Ok(white) = parse_hex("#FFF")
white |> darken(50) |> alpha(90) |> to_rgba() // and here return String
```

Maybe use it like this:

```gleam
fn primary() -> String {
  parse_hex("#00F")
	|> result.map(darken(_, 50))
	|> result.map(alpha(_, 90))
	|> result.map(to_rgba)
	|> result.replace_error("blue")
	|> result.unwrap_both()
}
```

Or: 

```gleam
fn primary() -> Result(String, LumiereError) {
  parse_hex("#00F")
	|> result.map(darken(_, 50))
	|> result.map(alpha(_, 90))
	|> result.map(to_rgba)
}

fn button() -> Class {
  let assert Ok(primary_color) = primary()
  sketch.class([
    sketch.background(primary_color),
  ])
}
```

You can ignore the `Result(Color, Color)`, for now. It's just me prototyping stuff 😄. 

I'm focusing now on the convert/parse functions so I'll get back to the final API later.

Thinking about all of these, what do you suggest?

**Comment by ghivert** [comment]
> ```gleam
> let assert Ok(white) = parse_hex("#FFF")
> white |> darken(50) |> alpha(90) |> to_rgba() // and here return String
> ```

This is exactly how I'll do when I'm 100% sure it won't throw.

---

This is how I would handle colors in my app.

```gleam
// colors.gleam
import lumiere.{type Color}

pub type Palette {
  Palette(white: Color, black: Color)
}

pub fn palette() {
  let assert Ok(white) = lumiere.parse("#FFF")
  let assert Ok(black) = lumiere.parse("#000")
  Palette(white, black)
}
```

---

You should also keep in mind you're in Gleam, so you have access to `use`.

```gleam
fn primary() -> String {
  let color = {
    use color <- result.map(parse("#00F"))
    color
    |> lumiere.darken(50)
    |> lumiere.alpha(90)
    |> lumiere.to_rgba()
  }
  color |> result.unwrap("blue")
}
```

You could also define your function for sketch.

```gleam
fn color(color: Result(lumiere.Color, error)) -> sketch.Style(media, pseudo) {
  result.map(color, fn (c) { sketch.color(lumiere.to_rgb(c)) })
  |> result.unwrap(sketch.none())
}
```

Like this, you don't care about the Result most of the time.

--- 

I think it's better to let the devs handle the Result on their own, and keep it unopinionated. You could always build on top and add your layer managing the Result easily, while the reverse is not possible (when the foundations are opinionated, it's harder to build on top if you disagree on some points).

**Comment by brunoti** [comment]
Awesome. I'm convinced!
_____________________________________________________________

```gleam
fn color(color: Result(lumiere.Color, error)) -> sketch.Style(media, pseudo) {
  result.map(color, fn (c) { sketch.color(lumiere.to_rgb(c)) })
  |> result.unwrap(sketch.none())
}
```
I have been thinking of something like this since day one!
